clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -6196,7 +6283,16 @@ void SYCLIntegrationHeader::emit(raw_ostream &O) {
   O << "#include <sycl/detail/defines_elementary.hpp>\n";
   O << "#include <sycl/detail/kernel_desc.hpp>\n";
   O << "#include <sycl/ext/oneapi/experimental/free_function_traits.hpp>\n";
-
+  // When using work group memory parameters in free kernel functions, the
```

Comment:
For the work group memory type, the integration header emits a forward declaration of the form 
```
template<typename, typename>
class work_group_memory;
```
The second parameter actually has a default argument as per the spec, but clearly it is not included here(I guess to avoid redefinition of default argument which would be a compiler error).
So when you use a type such as `work_group_memory<int>` as a free function kernel parameter, the compiler complains because it thinks it should have two parameters and not just one.

The file `sycl/ext/oneapi/experimental/work_group_memory_forward_decl.hpp`, as the name suggests, contains the forward declaration of `work_group_memory` where the default argument is also included. So I just manually include this file into the integration header.  Unfortunately, this fixed my initial issue but it has broken 2 tests in Windows pre-commit CI(they are now in progress since I made a change but you can look at the CI results for the previous commit).

Reply:
Windows pre-commit fails: https://github.com/intel/llvm/actions/runs/11522402935/job/32081852120?pr=15861

Reply:
To follow up on this, I'm not quite how this issue has not come up yet. It seems to me that in general, every templated class with a default argument should be problematic to use as a free function kernel parameter but it looks like current tests are not catching this.

Reply:
The implementation of free function kernels is not complete, and I know there are many limitations currently.  My guess is that no one has added support for parameters that are templated types, and no one tested this yet.  :-(

Reply:
> For the work group memory type, the integration header emits a forward declaration of the form
> 
> ```
> template<typename, typename>
> class work_group_memory;
> ```
> 
> The second parameter actually has a default argument as per the spec, but clearly it is not included here(I guess to avoid redefinition of default argument which would be a compiler error). So when you use a type such as `work_group_memory<int>` as a free function kernel parameter, the compiler complains because it thinks it should have two parameters and not just one.
> 
> The file `sycl/ext/oneapi/experimental/work_group_memory_forward_decl.hpp`, as the name suggests, contains the forward declaration of `work_group_memory` where the default argument is also included. So I just manually include this file into the integration header. Unfortunately, this fixed my initial issue but it has broken 2 tests in Windows pre-commit CI. This solution may be a hack, so perhaps the user you tagged(@Fznamznon ) may have a better solution in mind.

Also tagging @steffenlarsen , @AlexeySachkov for awareness.

Reply:
Hmm... I am not sure what a possible solution here is. I also don't think this is restricted to free functions. Won't any type we forward declare for kernel name with default arguments run into a similar issue? Tagging @AaronBallman, @tahonermann and @premanandrao as well for any thoughts/ideas.

>>The file sycl/ext/oneapi/experimental/work_group_memory_forward_decl.hpp, as the name suggests, contains the forward declaration of work_group_memory where the default argument is also included.

I assume this could cause some sort of ambiguity error maybe if compiler does not know how to resolve between different declarations.

Reply:
I was just talking to @elizabethandrews about this and the only solution that I could think of was perhaps include the forward declaration header file itself (like the one above) when something that we recognize (like work_group_memory) needs to be forward declared.

But this seems to be what you tried yourself.   What is the nature of the Windows CI tests breakage?

Reply:
>>the only solution that I could think of was perhaps include the forward declaration header file itself (like the one above) when something that we recognize (like work_group_memory) needs to be forward declared.

My concern here is we will now have to start inspecting argument types and add restrictions on what to forward declare and not based on the type of the argument. It feels like questionable design to me.

Reply:
> I was just talking to @elizabethandrews about this and the only solution that I could think of was perhaps include the forward declaration header file itself (like the one above) when something that we recognize (like work_group_memory) needs to be forward declared.
> 
> But this seems to be what you tried yourself. What is the nature of the Windows CI tests breakage?

Quite surprisingly, the Windows CI is now passing. Th initial failures(https://github.com/intel/llvm/actions/runs/11522402935/job/32081852120?pr=15861) did seem a bit unrelated to me so I suppose either the fails were unrelated or a later change fixed it and I mistakenly looked at an outdated CI result.

Reply:
>  It feels like questionable design to me.

Indeed.  I am not fond of it, but I can't think of another solution for this.   Maybe others we have tagged can.

Reply:
I'm not sure I'm following here. I wouldn't expect a forward declaration to be needed for the free function kernel itself; the original declarations visible during device compilation should suffice for that. The forward declarations in the integration header are only necessary to satisfy other declarations in the integration header. None of those declarations should have any reliance on default arguments. What am I missing? Can someone post an example of the (presumably generated) code that fails to compile?

Reply:
> I'm not sure I'm following here. I wouldn't expect a forward declaration to be needed for the free function kernel itself; the original declarations visible during device compilation should suffice for that. The forward declarations in the integration header are only necessary to satisfy other declarations in the integration header. None of those declarations should have any reliance on default arguments. What am I missing? Can someone post an example of the (presumably generated) code that fails to compile?

Indeed, after inspecting the integration header, I can confirm the current implementation of free function kernels does generate a forward declaration for the free function which in turn requires forward declarations of the argument types such as work group memory.

Reply:
Sure, but that forward declaration of the free function shouldn't have a dependency on default arguments. Can you paste an example from your integration header?

Reply:
> Sure, but that forward declaration of the free function shouldn't have a dependency on default arguments. Can you paste an example from your integration header?

Here is the source file:

```
#include <sycl/detail/core.hpp>
#include <sycl/ext/oneapi/free_function_queries.hpp>
#include <sycl/ext/oneapi/experimental/work_group_memory.hpp>

using namespace sycl;

queue q;
context ctx = q.get_context();

extern "C"
SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(
    (ext::oneapi::experimental::nd_range_kernel<1>))
void foo(
    sycl::ext::oneapi::experimental::work_group_memory<int> mem) {
mem = 42;
 }


int main() {
#ifndef __SYCL_DEVICE_ONLY__
  // Get the kernel object for the "mykernel" kernel.
  kernel_bundle Bundle = get_kernel_bundle<bundle_state::executable>(ctx);
  kernel_id sum_id = ext::oneapi::experimental::get_kernel_id<foo>();
  kernel k_sum = Bundle.get_kernel(sum_id);
  q.submit([&](sycl::handler &cgh) {
     ext::oneapi::experimental::work_group_memory<int> mem{cgh};
     cgh.set_args(mem);
     cgh.parallel_for(nd_range{ {1}, {1} }, k_sum);
   }).wait();
#endif
return 0;
}
```

Here is the produces integration header:

```
// This is auto-generated SYCL integration header.

#include <sycl/detail/defines_elementary.hpp>
#include <sycl/detail/kernel_desc.hpp>
#include <sycl/ext/oneapi/experimental/free_function_traits.hpp>

#ifndef SYCL_LANGUAGE_VERSION
#define SYCL_LANGUAGE_VERSION 202001
#endif //SYCL_LANGUAGE_VERSION

// Forward declarations of templated kernel function types:

namespace sycl {
inline namespace _V1 {
namespace detail {
// names of all kernels defined in the corresponding source
static constexpr
const char* const kernel_names[] = {
  "__sycl_kernel_foo",
  "",
};

// array representing signatures of all kernels defined in the
// corresponding source
static constexpr
const kernel_param_desc_t kernel_signatures[] = {
  //--- __sycl_kernel_foo
  { kernel_param_kind_t::kind_work_group_memory, 16, 0 },

  { kernel_param_kind_t::kind_invalid, -987654321, -987654321 },
};

// Specializations of KernelInfo for kernel function types:
template <> struct KernelInfoData<'_', '_', 's', 'y', 'c', 'l', '_', 'k', 'e', 'r', 'n', 'e', 'l', '_', 'f', 'o', 'o'> {
  __SYCL_DLL_LOCAL
  static constexpr const char* getName() { return "__sycl_kernel_foo"; }
  __SYCL_DLL_LOCAL
  static constexpr unsigned getNumParams() { return 1; }
  __SYCL_DLL_LOCAL
  static constexpr const kernel_param_desc_t& getParamDesc(unsigned i) {
    return kernel_signatures[i+0];
  }
  __SYCL_DLL_LOCAL
  static constexpr bool isESIMD() { return 0; }
  __SYCL_DLL_LOCAL
  static constexpr const char* getFileName() {
#ifndef NDEBUG
    return "free_function.cpp";
#else
    return "";
#endif
  }
  __SYCL_DLL_LOCAL
  static constexpr const char* getFunctionName() {
#ifndef NDEBUG
    return "void (class sycl::ext::oneapi::experimental::work_group_memory<int>)";
#else
    return "";
#endif
  }
  __SYCL_DLL_LOCAL
  static constexpr unsigned getLineNumber() {
#ifndef NDEBUG
    return 13;
#else
    return 0;
#endif
  }
  __SYCL_DLL_LOCAL
  static constexpr unsigned getColumnNumber() {
#ifndef NDEBUG
    return 6;
#else
    return 0;
#endif
  }
  // Returns the size of the kernel object in bytes.
  __SYCL_DLL_LOCAL
  static constexpr long getKernelSize() { return 0; }
};

} // namespace detail
} // namespace _V1
} // namespace sycl

// Definition of __sycl_kernel_foo as a free function kernel

// Forward declarations of kernel and its argument types:
namespace std {
template <typename ..._Elements> class tuple;
}
namespace sycl { inline namespace _V1 { namespace ext { namespace oneapi { namespace experimental {
template <typename PropertiesT> class properties;
}}}}}
namespace sycl { inline namespace _V1 { namespace ext { namespace oneapi { namespace experimental {
template <typename DataT, typename PropertyListT> class work_group_memory;
}}}}}

extern "C" void foo(sycl::ext::oneapi::experimental::work_group_memory<int> mem);
static constexpr auto __sycl_shim1() {
  return (void (*)(class sycl::ext::oneapi::experimental::work_group_memory<int>))foo;
}
namespace sycl {
template <>
struct ext::oneapi::experimental::is_kernel<__sycl_shim1()> {
  static constexpr bool value = true;
};
template <>
struct ext::oneapi::experimental::is_nd_range_kernel<__sycl_shim1(), 1> {
  static constexpr bool value = true;
};
}

#include <sycl/kernel_bundle.hpp>

// Definition of kernel_id of __sycl_kernel_foo
namespace sycl {
template <>
kernel_id ext::oneapi::experimental::get_kernel_id<__sycl_shim1()>() {
  return sycl::detail::get_kernel_id_impl(std::string_view{"__sycl_kernel_foo"});
}
}
```

Reply:
Thank you. For ease of reference, here is how `work_group_memory` is declared in `work_group_memory_forward_decl.hpp`:
```
template <typename DataT, typename PropertiesT = int>
class work_group_memory;
```

This indeed looks like a bug in the integration header generation. The definition of `__sycl_shim1` should be declared as follows (with the `PropertiesT` default argument `int` inserted as the second template argument of `work_group_memory` in both locations):
```
extern "C" void foo(sycl::ext::oneapi::experimental::work_group_memory<int, int> mem);
static constexpr auto __sycl_shim1() {
  return (void (*)(class sycl::ext::oneapi::experimental::work_group_memory<int, int>))foo;
}
```

Reply:
> Thank you. For ease of reference, here is how `work_group_memory` is declared in `work_group_memory_forward_decl.hpp`:
> 
> ```
> template <typename DataT, typename PropertiesT = int>
> class work_group_memory;
> ```
> 
> This indeed looks like a bug in the integration header generation. The definition of `__sycl_shim1` should be declared as follows (with the `PropertiesT` default argument `int` inserted as the second template argument of `work_group_memory` in both locations):
> 
> ```
> extern "C" void foo(sycl::ext::oneapi::experimental::work_group_memory<int, int> mem);
> static constexpr auto __sycl_shim1() {
>   return (void (*)(class sycl::ext::oneapi::experimental::work_group_memory<int, int>))foo;
> }
> ```

Oops, looks like you referenced the `work_group_memory_forward_decl` from the dummy clang/test directory. That is purely for testing purposes so the value of the default argument does not matter so I just put int.
Further down in the list of changed files should be the real `work_group_memory_forward_decl`

Reply:
> Thank you. For ease of reference, here is how `work_group_memory` is declared in `work_group_memory_forward_decl.hpp`:
> 
> ```
> template <typename DataT, typename PropertiesT = int>
> class work_group_memory;
> ```
> 
> This indeed looks like a bug in the integration header generation. The definition of `__sycl_shim1` should be declared as follows (with the `PropertiesT` default argument `int` inserted as the second template argument of `work_group_memory` in both locations):
> 
> ```
> extern "C" void foo(sycl::ext::oneapi::experimental::work_group_memory<int, int> mem);
> static constexpr auto __sycl_shim1() {
>   return (void (*)(class sycl::ext::oneapi::experimental::work_group_memory<int, int>))foo;
> }
> ```

Would the frontend team be willing to open a PR soon to correct this assuming its not too much work?

Reply:
> This indeed looks like a bug in the integration header generation. The definition of `__sycl_shim1` should be declared as follows (with the `PropertiesT` default argument `int` inserted as the second template argument of `work_group_memory` in both locations):

I think this is not quite right.  The second template parameter should not be `int`.  It should be `empty_properties_t` aka `properties<std::tuple<>>`.  See my comment above.

Reply:
> > This indeed looks like a bug in the integration header generation. The definition of `__sycl_shim1` should be declared as follows (with the `PropertiesT` default argument `int` inserted as the second template argument of `work_group_memory` in both locations):
> 
> I think this is not quite right. The second template parameter should not be `int`. It should be `empty_properties_t` aka `properties<std::tuple<>>`. See my comment above.

Yeah, @tahonermann was misled by the dummy `work_group_memory_forward_decl.hpp` that I had to create for clang testing purposes. Now that we've decided to go a different route, that file is actually not needed at all so I'll just remove it to avoid confusion.

Reply:
Oops, I neglected to check that my quick search actually found the thing I thought I was searching for :)

That doesn't change the analysis though. @lbushi25, I recommend you open a Jira issue for the FE team. I suspect a fix won't be too difficult, but I'm not very familiar with that code either.